---
output:
  flexdashboard::flex_dashboard:
    orientation: rows
    theme: cosmo
params: 
  patient_dir: 
    value: "results"
  fusion_method:
    value: "arriba"
  set_title:
    value: "Patient Report"
  snv_caller:
    value: NULL
editor_options: 
  chunk_output_type: console
---
---
title: `r params$set_title`
---

<style>
/* Fix DT / Flexdashboard container cut off */
.chart-stage-flex {
  overflow: scroll !important;
}
</style>

```{r set_root_dir}
# directories
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
```

```{r load_libraries}
source(file.path(root_dir, "code", "utils", "load_libraries.R"))
```

```{r set_options, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(comment = NA)
`%>%` <- dplyr::`%>%`
```

```{r set_params}
patient_dir <- params$patient_dir
fusion_method <- params$fusion_method
snv_caller <- params$snv_caller
```

```{r call_driver, echo=FALSE, message=FALSE, results="hide"}
# function for viewing datatables
source(file.path(root_dir, "code", "utils", "view_dt.R"))
# call driver which calls rest of the scripts to generate outputs
source(file.path(root_dir, "code", "driver.R"))
```

Key Clinical Findings (P1) {data-navmenu="Key Findings"}
===
Row {data-height=200}
---   
### {data-width=200}

```{r d3b_logo, echo=FALSE, out.width = "100%"}
knitr::include_graphics(file.path(root_dir, "figures", "d3b_logo.png"))
```
    
### Patient & Sample Information

```{r patient_info}
x = kable(patient_sample_info(), format="html") %>%
  kable_styling(bootstrap_options = c("bordered", "condensed", "responsive")) %>% 
  column_spec(c(1,3), bold=T, border_left = "2px solid gray", border_right = "2px solid gray")
gsub("<thead>.*</thead>", "", x)
```

Row {data-height=500}
---
    
### Key Clinical Findings
    
```{r key_findings}
if(nrow(key_clinical_findings_output) >= 1){
  view_dt_minimal(key_clinical_findings_output)
}
```

Row {data-height=300}
---
    
### Disease Specific Information
    
```{r disease_specific_info}
if(nrow(disease_specific_information_output) >= 1){
  DT::datatable(disease_specific_information_output, 
                extensions = "FixedColumns",
                class = "cell-border stripe",
                options = list(dom = "t",
                               scrollX = TRUE,
                               fixedColumns = TRUE), 
                rownames = F, colnames = c("Type", "Value"))
}
```

### Cancer Germline Variants

```{r filtered_germline_vars}
if(nrow(filtered_germ_vars) >= 1){
  DT::datatable(filtered_germ_vars, 
                extensions = "FixedColumns",
                class = "cell-border stripe",
                options = list(dom = "t",
                               scrollX = TRUE,
                               fixedColumns = TRUE), 
                rownames = F)
}
```

### Genomic Summary
    
```{r genomic_summary}
if(nrow(genomic_summary_output) >= 1){
  DT::datatable(genomic_summary_output, 
                extensions = "FixedColumns",
                class = "cell-border stripe",
                options = list(dom = "t",
                               scrollX = TRUE,
                               fixedColumns = TRUE), 
                rownames = F, colnames = c("Type", "Value"))
}
```

All Findings (P1) {data-navmenu="Key Findings"}
===
Row {data-height=700}
---

### All Findings
    
```{r all_findings}
if(nrow(all_findings_output) >= 1){
  view_dt_minimal(all_findings_output)
}
```

Gene Alterations (P2) {data-navmenu="Transcriptomic Alterations"}
===
Row {data-height700}
---

### Gene Alterations 
    
```{r plot_genes, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "rnaseq_analysis", 
                   "diffexpr_genes_barplot_output.pdf")
if(file.exists(fname)){
 knitr::include_graphics(fname) 
}
```

Pathway Alterations: (P2) {data-navmenu="Transcriptomic Alterations"}
===
Row {data-height=700}
---

### Pathway Alterations

```{r plot_pathways, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "rnaseq_analysis", 
                   "diffreg_pathways_barplot_output.pdf")
if(file.exists(fname)){
 knitr::include_graphics(fname) 
}
```

Network Alterations: PBTA HGAT (P2) {data-navmenu="Transcriptomic Alterations"}
===
Row {data-height=700}
---

### Pathway Alterations (PBTA)

```{r network_alt_pbta_hgg, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "drug_recommendations", 
                   "ora_plots.pdf")
if(file.exists(fname)){
 knitr::include_graphics(fname) 
}
```

Tumor Burden/Signatures (P3)
===
Row {data-height=700}
---

### Tumor Mutational Signatures

```{r tm_signature}
if(exists("tumor_signature_output")){
  ggplotly(tumor_signature_output)
}
```

### Tumor Mutational Burden

```{r tm_burden, fig.width = 11, fig.height = 7}
if(exists("tmb_profile_output")){
  ggplotly(tmb_profile_output)
}
```

Immune Profile: Pediatric (P4) {data-navmenu="Immune Analyses"}
===
Row {data-height=700}
---

### Immune Profile (with PBTA)

```{r immune_profile_pbta, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "immune_analysis", 
                   "immune_scores_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Immune Profile: Adult (P4) {data-navmenu="Immune Analyses"}
===
Row {data-height=700}
---

### Immune Profile (with TCGA GBM)

```{r immune_profile_tcga, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "immune_analysis", 
                   "immune_scores_adult.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Tumor Inflammation Signature (P4) {data-navmenu="Immune Analyses"}
===
Row {data-height=700}
---

### Tumor Inflammation Signature Profile (TCGA + PBTA)

```{r tis_profile, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "immune_analysis", 
                   "tis_scores.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Transcriptomically Similar Samples (PBTA) (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=700}
---

### Nearest-Neighbor Analysis

```{r dimred_plot_pbta, fig.width = 14, fig.height = 10}
if(exists("dim_reduction_plot_pediatric")){
  ggplotly(dim_reduction_plot_pediatric, tooltip = "text")
}
```

### Transcriptomically Similar Patients

```{r similar_patients_pbta}
if(exists("transciptomically_similar_pediatric")){
  view_dt_minimal(transciptomically_similar_pediatric)
}
```

Survival (PBTA HGAT) (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=700}
---

### Survival (PBTA HGAT)

```{r km_plot_pbta, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "survival_analysis", 
                   "kaplan_meier_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname)
}
```

Immune Analyses (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=700}
---

### Immune Profile (Transcriptomically Similar Patients)

```{r immune_profile_gs, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "immune_analysis", 
                   "immune_scores_topcor_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname)
}
```

ssGSEA (PBTA HGAT) (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=700}
---

### ssGSEA (PBTA HGAT samples)

```{r ssgsea_hgg, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "ssgsea_scores_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Shared pathways (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=500}
---

### Shared pathways 

```{r shared_pathways_pediatric, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis", 
                   "pathway_analysis_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Shared Mutations (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=500}
---

### Shared Mutations (Transcriptomically Similar Patients)

```{r shared_mutations_pediatric, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "mutational_shared_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Shared Mutations AA changes (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=500}
---

### Shared Mutations AA changes (Transcriptomically Similar Patients)

```{r lollipop_shared_pediatric, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "lollipop_shared_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Recurrent Mutations (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=500}
---

### Recurrent Mutations (Transcriptomically Similar Patients)

```{r recurrent_mutations_pediatric, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "mutational_recurrent_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Recurrent Mutations AA changes (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=500}
---

### Recurrent Mutations AA changes (Transcriptomically Similar Patients)

```{r lollipop_recurrent_pediatric, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "lollipop_recurrent_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Shared CNV (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=500}
---

### Shared CNV (Transcriptomically Similar Patients)

```{r shared_cnv_pediatric, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "mutational_cnv_shared_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Recurrent CNV (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=500}
---

### Recurrent CNV (Transcriptomically Similar Patients)

```{r recurrent_cnv_pediatric, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "mutational_cnv_recurrent_pediatric.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```


Transcriptomically Similar Samples (TCGA) (P6) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=700}
---

### Nearest-Neighbor Analysis

```{r dimred_plot_tcga, fig.width = 14, fig.height = 10}
if(exists("dim_reduction_plot_adult")){
  ggplotly(dim_reduction_plot_adult, tooltip = "text")
}
```

### Transcriptomically Similar Patients

```{r similar_patients_tcga}
if(exists("transciptomically_similar_adult")){
  view_dt_minimal(transciptomically_similar_adult)
}
```

Survival (TCGA) (6) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=700}
---

### Survival (TCGA)

```{r km_plot_tcga, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "survival_analysis", 
                   "kaplan_meier_adult.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname)
}
```

Shared pathways (P6) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=500}
---

### Shared pathways 

```{r shared_pathways_adult, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "pathway_analysis_adult.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Shared Mutations (P6) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=500}
---

### Shared Mutations (Transcriptomically Similar Patients)

```{r shared_mutations_adult, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "mutational_shared_adult.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Shared Mutations AA changes (P6) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=500}
---

### Shared Mutations AA changes (Transcriptomically Similar Patients)

```{r lollipop_shared_adult, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "lollipop_shared_adult.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Recurrent Mutations (P6) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=500}
---

### Recurrent Mutations (Transcriptomically Similar Patients)

```{r recurrent_mutations_adult, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "mutational_recurrent_adult.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Recurrent Mutations AA changes (P6) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=500}
---

### Recurrent Mutations AA changes (Transcriptomically Similar Patients)

```{r lollipop_recurrent_adult, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "lollipop_recurrent_adult.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Shared CNV (P6) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=500}
---

### Shared CNV (Transcriptomically Similar Patients)

```{r shared_cnv_adult, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "mutational_cnv_shared_adult.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Recurrent CNV (P6) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=500}
---

### Recurrent CNV (Transcriptomically Similar Patients)

```{r recurrent_cnv_adult, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "transcriptomically_similar_analysis",
                   "mutational_cnv_recurrent_adult.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname) 
}
```

Circos Plot (P7) {data-navmenu="Genomic Landscape"}
===
Row {data-height=700}
---

### Circos

```{r circos, fig.width = 7, fig.height = 6}
fname <- file.path(patient_dir, "output",
                   "genomic_landscape_plots",
                   "circos_plot.png")
if(file.exists(fname)){
 knitr::include_graphics(fname)
}
```

Network View (P7) {data-navmenu="Genomic Landscape"}
===
Row {data-height=700}
---

### Network View

```{r plot_network, fig.width = 8, fig.height = 6}
if(exists("network_plot_output")){
  network_plot_output
}
```

CNA View (P7) {data-navmenu="Genomic Landscape"}
===
Row {data-height=700}
---

### CNA View

```{r plot_cnv, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- list.files(path = file.path(patient_dir, "copy-number-variations"), pattern = ".pdf", full.names = T)
if(file.exists(fname)){
 knitr::include_graphics(fname) 
}
```

<!-- PHGG: Expression and CNV Heatmap  (P8) {data-navmenu="Heatmaps"} -->
<!-- === -->
<!-- Row {data-height=700} -->
<!-- --- -->

<!-- ### Expression-CNV Heatmap (PHGG) -->

<!-- ```{r heatmap_phgg, echo=FALSE} -->
<!-- fname <- file.path(patient_dir, "output", "complexheatmap_phgg.png") -->
<!-- if(file.exists(fname)){ -->
<!--   knitr::include_graphics(fname)   -->
<!-- } -->
<!-- ``` -->

<!-- CGS: Expression and CNV Heatmap (P8) {data-navmenu="Heatmaps"} -->
<!-- === -->
<!-- Row -->
<!-- --- -->

<!-- ### Expression-CNV Heatmap (CGS) -->

<!-- ```{r heatmap_cgs, echo=FALSE} -->
<!-- fname <- file.path(patient_dir, "output", "complexheatmap_cgs.png") -->
<!-- if(file.exists(fname)){ -->
<!--   knitr::include_graphics(fname) -->
<!-- } -->
<!-- ``` -->

Oncogrid (HGG-DMG)  (P9)
===
Row {data-height=700}
---

### Oncogrid (HGG-DMG)

```{r oncogrid_hgg_dmg, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "oncogrid_analysis",
                   "complexheatmap_oncogrid.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname)
}
```

Targeted findings (P10) {data-navmenu="Drug Recommendations"}
===
Row {data-height=700}
---

### Drug Recommendations

```{r targeted_findings}
if(exists("oncokb_output")){
  view_dt_minimal(oncokb_output)
}
```

DGE (P10) {data-navmenu="Drug Recommendations"}
===
Row {data-height=700}
---

### DGE

```{r transcriptome_drug_rec}
if(exists("transcriptome_drug_rec_output")){
  view_dt_minimal(transcriptome_drug_rec_output)
}
```

DGE Plots (P10) {data-navmenu="Drug Recommendations"}
===
Row {data-height=700}
---

### DGE Plots

```{r dge_plots, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "drug_recommendations",
                   "drug_dge_density_plots", 
                   "top_drug_dge_density_plots.pdf")
if(file.exists(fname)){
 knitr::include_graphics(fname)
}
```

Drug Signatures (P10) {data-navmenu="Drug Recommendations"}
===
Row {data-height=700}
---

### Drug Signatures 

```{r drug_signatures, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "drug_recommendations",
                   "drug_pathways_barplot.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname)
}
```

Drug-Gene Network: GTEx Brain (P10) {data-navmenu="Drug Recommendations"}
===
Row {data-height=700}
---

### GTEx Brain

```{r drug_gene_network_gtex, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "drug_recommendations",
                   "GTExBrain_dsea_go_mf_output.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname)
}
```

Drug-Gene Network: PBTA HGAT (P10) {data-navmenu="Drug Recommendations"}
===
Row {data-height=700}
---

### PBTA HGAT

```{r drug_gene_network_pbta_hgg, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "drug_recommendations",
                   "PBTA_HGG_dsea_go_mf_output.pdf")
if(file.exists(fname)){
 knitr::include_graphics(fname) 
}
```

Drug-Gene Network: PBTA (P10) {data-navmenu="Drug Recommendations"}
===
Row {data-height=700}
---

### PBTA

```{r drug_gene_network_pbta, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "drug_recommendations",
                   "PBTA_ALL_dsea_go_mf_output.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname)
}
```

Synergistic Drug Pairs (P10) {data-navmenu="Drug Recommendations"}
===
Row {data-height=700}
---

### Synergistic Drug Pairs

```{r synergistic_drug_pairs, echo=FALSE, out.width = "100%", out.height="100%"}
fname <- file.path(patient_dir, "output", 
                   "drug_synergy", 
                   "combined_qSig_synergy_score_top10.pdf")
if(file.exists(fname)){
  knitr::include_graphics(fname)
}
```