---
title: "Adult Mutational MAF Analysis Graph Conversion"
author: "Run Jin"
date: "8/23/2021"
output: html_document
---
```{r setup, include=FALSE}
library(dplyr)
# BiocManager::install("maftools", force =TRUE)
library(maftools)
library(TCGAbiolinks)

pnoc008_sample_of_interest = "PNOC008-36"
```

Define directory
```{r define directory, include=FALSE}
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")

adult_dir <- file.path(root_dir, "graph-updates/plots/adult")
if (!dir.exists(adult_dir )) {
  dir.create(adult_dir , recursive = TRUE)
}

lollipop_dir <- file.path(adult_dir, "lollipop")
if (!dir.exists(lollipop_dir )) {
  dir.create(lollipop_dir , recursive = TRUE)
}

plotmaf_dir <- file.path(adult_dir, "plotmaf")
if (!dir.exists(plotmaf_dir )) {
  dir.create(plotmaf_dir , recursive = TRUE)
}

oncoplot_dir <- file.path(adult_dir, "oncoplot")
if (!dir.exists(oncoplot_dir )) {
  dir.create(oncoplot_dir , recursive = TRUE)
}

```


## P6 Adult Tumor Analysis - Mutational Analysis
### Query for TCGA MAF file and filter using histology file
```{r clean up TCGA MAF file}
# Query with TCGABiolinks
query <- GDCquery(project = "TCGA-GBM", 
                  data.category = "Simple Nucleotide Variation", 
                  access="open", 
                  legacy = F, 
                  data.type = "Masked Somatic Mutation", 
                  workflow.type = "MuTect2 Variant Aggregation and Masking")

GDCdownload(query)
tcga_maf <- GDCprepare(query, add.gistic2.mut = T)

tcga_maf$Tumor_Sample_Barcode <- gsub('D.*|W.*', '',tcga_maf$Tumor_Sample_Barcode)

# Filter to the samples that we have RNA information 
tcga_clin <- readRDS(file.path(data_dir, "tcga_gbm_clinical.rds")) %>%
  dplyr::rename(Tumor_Sample_Barcode = sample_barcode)

tcga_maf <- tcga_maf %>%
  filter(Tumor_Sample_Barcode %in% tcga_clin$Tumor_Sample_Barcode)

```

### Read in MAF for PNOC008
```{r read in PNOC008 MAF}
consensus_maf_pnoc008 <- readRDS(file.path(data_dir, "pnoc008_consensus_maf_combined.rds"))
```

### Read in mutational analysis
```{r read in for mutational analysis}
mutational_analysis <- readRDS(file.path(data_dir, "mutational_analysis_adult.rds"))
```

### Genearte figures for recurrent mutations 
#### Prepare combined MAF for recurrent mutations
```{r prepare recurrent MAF}
recurrent_subjects <- mutational_analysis$recurrent_alterations %>% 
  filter(Alteration_Datatype == "Mutation") %>% 
  pull(subject_id) %>% unique() %>%
  append(pnoc008_sample_of_interest)

recurrent_alterations_pnoc008 <- recurrent_subjects[recurrent_subjects %in% consensus_maf_pnoc008$subjectID] 
recurrent_alterations_tcga <- recurrent_subjects[!recurrent_subjects %in% consensus_maf_pnoc008$subjectID]

stopifnot(recurrent_alterations_tcga %in% tcga_maf$Tumor_Sample_Barcode)

# Use the kids_first_biospecimen_id to filter TCGA combined consensus SNV
matched_maf_tcga_recurrent <- tcga_maf %>% filter(Tumor_Sample_Barcode %in% recurrent_alterations_tcga)

# filter based on subjectID
matched_maf_pnoc008_recurrent <- consensus_maf_pnoc008 %>% filter(subjectID %in% recurrent_alterations_pnoc008)

# combine the maf to prepare for MAF object
common_columns <- intersect(colnames(tcga_maf), colnames(consensus_maf_pnoc008))

matched_maf_tcga_recurrent <- matched_maf_tcga_recurrent %>% select(common_columns)
matched_maf_pnoc008_recurrent <- matched_maf_pnoc008_recurrent %>% select(common_columns)

combined_maf_recurrent <- rbind(matched_maf_tcga_recurrent, matched_maf_pnoc008_recurrent) %>%
  rename(AAChange = HGVSp_Short) 

```

#### Prepare combined histology for recurrent mutations
```{r prepare recurrent histology}
match_pnoc008 <- consensus_maf_pnoc008 %>% 
  select(subjectID, Tumor_Sample_Barcode) %>% distinct() 

histology_pnoc008 <- readRDS(file.path(data_dir, "pnoc008_clinical.rds")) %>%
  left_join(match_pnoc008) %>%
  rename(
    gender = sex,
    age_at_diagnosis_in_days = age_diagnosis_days
  ) 

histology_pnoc008_recurrent <- histology_pnoc008 %>% 
  filter(subjectID %in% recurrent_alterations_pnoc008) 

tcga_clin_recurrent <- tcga_clin %>% 
  filter(Tumor_Sample_Barcode %in% recurrent_alterations_tcga)

common_columns_histology <- intersect(colnames(histology_pnoc008), colnames(tcga_clin))

histology_pnoc008_recurrent <- histology_pnoc008_recurrent %>% select(common_columns_histology)
tcga_clin_recurrent <- tcga_clin_recurrent%>% select(common_columns_histology)

combined_histology_recurrent <- rbind(histology_pnoc008_recurrent, tcga_clin_recurrent)

```

#### Create MAF object
```{r prepare recurrent MAF object}
maf_object_recurrent <- read.maf(maf = combined_maf_recurrent , clinicalData = combined_histology_recurrent)
maf_object_recurrent
```

#### Plot MAF object
```{r plot MAF summary adult recurrent}
pdf(file.path(plotmaf_dir, "mutational_recurrent_adult.pdf"))
plotmafSummary(maf = maf_object_recurrent, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
dev.off()

```
#### Generate lollipop plot 
```{r plot lollipop plot adult recurrent}
# First find the ten most mutated gene
summary_genes <- getGeneSummary(maf_object_recurrent) %>% 
  arrange(desc(AlteredSamples)) 

# find the most mutated 10 genes - PAML2-AKAP2 does not have structure available 
top10_gene_list <- summary_genes[1:10, ] %>% 
  pull(Hugo_Symbol) %>% unique()

top10_gene_list_position <- lapply(top10_gene_list, function(x) {
  protein_changes <- combined_maf_recurrent %>% filter(Hugo_Symbol == x) %>% 
    filter(AAChange != ".") %>% 
    pull(AAChange)
  positions <- gsub(".*?([0-9]+).*", "\\1", protein_changes)
})

names(top10_gene_list_position) <- top10_gene_list 

pdf(file.path(lollipop_dir, "lollipop_recurrent_adult.pdf"), onefile = TRUE)
lapply(top10_gene_list, function(x){
  lollipopPlot(
  maf = maf_object_recurrent,
  gene = x,
  showMutationRate = TRUE, 
  labelPos = top10_gene_list_position[[x]]
)
})
dev.off()

```

### Genearte figures for shared genes
#### Prepare combined MAF for shared mutations
```{r prepare shared MAF}
shared_subjects <- mutational_analysis$shared_genes %>% 
  filter(Alteration_Datatype == "Mutation") %>% 
  pull(subject_id) %>% unique() %>%
  append(pnoc008_sample_of_interest)

shared_subjects_pnoc008 <- shared_subjects[shared_subjects %in% consensus_maf_pnoc008$subjectID]
shared_subjects_tcga <- shared_subjects[!shared_subjects %in% consensus_maf_pnoc008$subjectID]

# stopifnot(shared_subjects_tcga%in% tcga_maf$Tumor_Sample_Barcode)

# Use the kids_first_biospecimen_id to filter TCGA combined consensus SNV
matched_maf_tcga_shared <- tcga_maf %>% filter(Tumor_Sample_Barcode %in% shared_subjects_tcga)

# filter based on subjectID
matched_maf_pnoc008_shared <- consensus_maf_pnoc008 %>% filter(subjectID %in% shared_subjects_pnoc008)

# combine the maf to prepare for MAF object
matched_maf_tcga_shared <- matched_maf_tcga_shared %>% select(common_columns)
matched_maf_pnoc008_shared <- matched_maf_pnoc008_shared %>% select(common_columns)

combined_maf_shared <- rbind(matched_maf_tcga_shared , matched_maf_pnoc008_shared) %>%
  rename(AAChange = HGVSp_Short) 

```

#### Prepare combined histology for recurrent mutations
```{r prepare shared histology}
histology_pnoc008_shared <- histology_pnoc008 %>% 
  filter(subjectID %in% shared_subjects_pnoc008) 

tcga_clin_shared <- tcga_clin %>% 
  filter(Tumor_Sample_Barcode %in% shared_subjects_tcga)

histology_pnoc008_shared <- histology_pnoc008_shared %>% select(common_columns_histology)
tcga_clin_shared <- tcga_clin_shared%>% select(common_columns_histology)

combined_histology_shared <- rbind(histology_pnoc008_shared, tcga_clin_shared)

```

#### Create MAF object
```{r prepare shared MAF object}
maf_object_shared <- read.maf(maf = combined_maf_shared , clinicalData = combined_histology_shared)
maf_object_shared
```

#### Plot MAF object
```{r plot MAF summary adult shared}
pdf(file.path(plotmaf_dir, "mutational_shared_adult.pdf"))
plotmafSummary(maf = maf_object_shared, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
dev.off()

```

#### Generate lollipop plot 
```{r plot lollipop plot adult shared}
# First find the ten most mutated gene
summary_genes <- getGeneSummary(maf_object_shared) %>% 
  arrange(desc(AlteredSamples)) 

# find the most mutated 10 genes 
top10_gene_list <- summary_genes[1:10, ] %>% 
  pull(Hugo_Symbol) %>% unique()

top10_gene_list_position <- lapply(top10_gene_list, function(x) {
  protein_changes <- combined_maf_shared %>% filter(Hugo_Symbol == x) %>% 
    filter(AAChange != ".") %>% 
    pull(AAChange)
  positions <- gsub(".*?([0-9]+).*", "\\1", protein_changes)
})

names(top10_gene_list_position) <- top10_gene_list 

pdf(file.path(lollipop_dir, "lollipop_shared_adult.pdf"), onefile = TRUE)
lapply(top10_gene_list, function(x){
  lollipopPlot(
  maf = maf_object_shared,
  gene = x,
  showMutationRate = TRUE, 
  labelPos = top10_gene_list_position[[x]]
)
})
dev.off()


```

### Genearte Oncoplot for CNV

#### Read in CNV files 
```{r read in CNV}
pnoc008_cnv <- readRDS(file.path(data_dir, "pnoc008_cnv_filtered.rds"))
tcga_cnv <- readRDS(file.path(data_dir, "tcga_gbm_cnv_filtered.rds"))
```

#### Prepare combined MAF for recurrent CNV
```{r prepare recurrent CNV MAF}
recurrent_subjects <- mutational_analysis$recurrent_alterations %>% 
  filter(Alteration_Datatype == "CNV") %>% 
  pull(subject_id) %>% unique() %>%
  append(pnoc008_sample_of_interest)

recurrent_subjects_pnoc008 <- recurrent_subjects[recurrent_subjects %in% consensus_maf_pnoc008$subjectID]
recurrent_subjects_tcga <- recurrent_subjects[!recurrent_subjects %in% consensus_maf_pnoc008$subjectID]

# Use the kids_first_biospecimen_id to filter TCGA combined consensus SNV
matched_maf_tcga_recurrent <- tcga_maf %>% filter(Tumor_Sample_Barcode %in% recurrent_subjects_tcga)

# filter based on subjectID
matched_maf_pnoc008_recurrent  <- consensus_maf_pnoc008 %>% filter(subjectID %in% recurrent_subjects_pnoc008)

# combine the maf to prepare for MAF object
matched_maf_tcga_recurrent <- matched_maf_tcga_recurrent %>% select(common_columns)
matched_maf_pnoc008_recurrent <- matched_maf_pnoc008_recurrent %>% select(common_columns)

combined_maf_recurrent <- rbind(matched_maf_tcga_recurrent, matched_maf_pnoc008_recurrent) %>%
  dplyr::rename(AAChange = HGVSp_Short) 

```


#### Filter CNV files to contain only recurrent specimens
```{r filter CNV to samples in recurrent CNV lists}
pnoc008_cnv_recurrent <- pnoc008_cnv %>% 
  filter(SampleID %in% recurrent_subjects_pnoc008 )
  
tcga_cnv_recurrent <- tcga_cnv %>% 
  filter(Kids_First_Biospecimen_ID %in% recurrent_subjects_tcga)

combined_cnv_recurrent <- rbind(pnoc008_cnv_recurrent, tcga_cnv_recurrent)
```

#### Clean up combined CNV recurrent 
```{r clean up combined recurrent cnv}
combined_cnv_recurrent <- combined_cnv_recurrent %>% 
  dplyr::mutate(CN = case_when(
    Alteration_Type == "Gain" ~ "ShallowAmp", 
    Alteration_Type == "Amplification" ~ "Amp", 
    Alteration_Type == "Loss" ~ "Del"
  )) %>% 
  dplyr::select(-Alteration_Type) %>%
  dplyr::rename(Sample_name = Kids_First_Biospecimen_ID) %>%
  dplyr::select(Gene, Sample_name, CN)

```

#### Generate MAF object with CNV data 
```{r generated MAF object with recurrent CNV}
maf_object_recurrent_cnv = read.maf(maf = combined_maf_recurrent, cnTable = combined_cnv_recurrent)

```

#### Plot Oncoplot 
```{r generated oncoplot for recurrent CNV}
pdf(file.path(oncoplot_dir, "mutational_cnv_recurrent_adult.pdf"))
oncoplot(maf = maf_object_recurrent_cnv, top=10)
dev.off()
```

#### Prepare combined MAF for shared CNV
```{r prepare shared CNV MAF}
shared_subjects <- mutational_analysis$shared_genes %>% 
  filter(Alteration_Datatype == "CNV") %>% 
  pull(subject_id) %>% unique() %>%
  append(pnoc008_sample_of_interest)

shared_subjects_pnoc008 <- shared_subjects[shared_subjects %in% consensus_maf_pnoc008$subjectID]
shared_subjects_tcga <- shared_subjects[!shared_subjects %in% consensus_maf_pnoc008$subjectID]

# Use the kids_first_biospecimen_id to filter TCGA combined consensus SNV
matched_maf_tcga_shared <- tcga_maf %>% filter(Tumor_Sample_Barcode %in% shared_subjects_tcga)

# filter based on subjectID
matched_maf_pnoc008_shared  <- consensus_maf_pnoc008 %>% filter(subjectID %in% shared_subjects_pnoc008)

# combine the maf to prepare for MAF object
matched_maf_tcga_shared <- matched_maf_tcga_shared %>% select(common_columns)
matched_maf_pnoc008_shared <- matched_maf_pnoc008_shared %>% select(common_columns)

combined_maf_shared <- rbind(matched_maf_tcga_shared, matched_maf_pnoc008_shared) %>%
  dplyr::rename(AAChange = HGVSp_Short) 

```

#### Filter CNV files to contain only recurrent specimens
```{r filter CNV to samples in shared CNV lists}
pnoc008_cnv_shared <- pnoc008_cnv %>% 
  filter(SampleID %in% shared_subjects_pnoc008 )
  
tcga_cnv_shared <- tcga_cnv %>% 
  filter(Kids_First_Biospecimen_ID %in% shared_subjects_tcga)

combined_cnv_shared <- rbind(pnoc008_cnv_shared, tcga_cnv_shared)
```

#### Clean up combined CNV in shared gene samples
```{r clean up combined shared cnv}
combined_cnv_shared <- combined_cnv_shared %>% 
  dplyr::mutate(CN = case_when(
    Alteration_Type == "Gain" ~ "ShallowAmp", 
    Alteration_Type == "Amplification" ~ "Amp", 
    Alteration_Type == "Loss" ~ "Del"
  )) %>% 
  dplyr::select(-Alteration_Type) %>%
  dplyr::rename(Sample_name = Kids_First_Biospecimen_ID) %>%
  dplyr::select(Gene, Sample_name, CN)

```

#### Generate MAF object with CNV data 
```{r generated MAF object with shared CNV}
maf_object_shared_cnv = read.maf(maf = combined_maf_shared, cnTable = combined_cnv_shared)

```

#### Plot Oncoplot 
```{r generated oncoplot for shared CNV}
pdf(file.path(oncoplot_dir, "mutational_cnv_shared_adult.pdf"))
oncoplot(maf = maf_object_shared_cnv, top=10)
dev.off()
```